File: P1087FBI树.py
class Node:
    def __init__(self, v):
        self.value = v
        self.left = None
        self.right = None

    def pastTraverse(self):
        if self.left != None:
            self.left.pastTraverse()
        if self.right != None:
            self.right.pastTraverse()
        print(self.value, end="")


def judge_string(ss):
    if not "1" in ss:
        return "B"
    if not "0" in ss:
        return "I"
    return "F"


def build_tree(org:str, root:Node)->None: 
    length = len(org)
    if length < 2:
        return
    lft = org[:length//2]
    rht = org[length//2:]
    root.left = Node(judge_string(lft))
    root.right = Node(judge_string(rht))
    if len(lft) > 1 and len(rht) > 1:
        build_tree(lft, root.left)
        build_tree(rht, root.right)


def main():
    gbg = input()  # 这个值可以不要
    del gbg
    org = input().strip()  # 这是输入进来的字符串
    root = Node(judge_string(org))
    build_tree(org, root)
    root.pastTraverse()

main()

File: P1087FBI树.cpp
#include <iostream>
#include <string>
using namespace std;

struct node
{
    string value;
    struct node *left;
    struct node *right;
    node(string s)
    {
        value = s;
        left = NULL;
        right = NULL;
    }
};

string judge_string(string tmp)
{
    if (tmp.size() > 0 && tmp.find("1") == string::npos)
    {
        return "B";
    }
    else if (tmp.size() > 0 && tmp.find("0") == string::npos)
    {
        return "I";
    }
    else
    {
        return "F";
    }
}

void build_tree(string org, node *root)
{
    // 字符串分割成左右两部分
    int length = org.size();
    if (length < 2)
    {
        return;
    }
    string lft = org.substr(0, length / 2);
    string rht = org.substr((length / 2), length / 2);
    // cout << "left: " << judge_string(lft) << " right: " << judge_string(rht) << endl;

    // 左右节点创建连接
    root->left = new node(judge_string(lft));
    root->right = new node(judge_string(rht));

    // 边界判定
    if (lft.size() > 1 && rht.size() > 1)
    {
        build_tree(lft, root->left);
        build_tree(rht, root->right);
    }
}

void pastTraverse(node *nd)
{
    // cout << nd->value;
    if (nd->left != NULL)
        pastTraverse(nd->left);
    if (nd->right != NULL)
        pastTraverse(nd->right);
    cout << nd->value;
}

int main()
{
    int n;
    cin >> n;
    string org;
    cin >> org;
    node *root = new node(judge_string(org));
    build_tree(org, root);
    pastTraverse(root);
    return 0;
}
